src/services.rs

```diff
@@ -0,0 +1,138 @@
+pub enum GitService {
+    GitHub,
+    SourceHut,
+}
+
+pub struct LineRange(pub usize, pub usize);
+
+impl LineRange {
+    fn linerange_for(&self, gs: &GitService) -> String {
+        match (gs, self.0, self.1)  {
+            (GitService::GitHub, a, b) if a == b => format!("#L{a}"),
+            (GitService::GitHub, a, b) => format!("#L{a}-{b}"),
+            /* SourceHut does not have multiline select at the time of writing. */
+            (GitService::SourceHut, a, _) => format!("#L{a}"),
+        }
+    }
+}
+
+impl std::fmt::Display for LineRange {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> Result<(), std::fmt::Error> {
+        match (self.0, self.1) {
+            (begin, end) if begin == end => write!(f, "#L{begin}"),
+            (begin, end) => write!(f, "#L{begin}-L{end}"),
+        }
+    }
+}
+
+// this is intended to build upon static strings.
+pub struct Data<'a> {
+    pub project: &'a str,
+    pub owner: &'a str,
+    pub path: &'a str,
+    pub branch_or_tag_name: Option<String>,
+    pub hash: Option<String>,
+    pub line_range: &'a Option<LineRange>,
+    pub service: GitService,
+}
+
+pub struct GitHub {}
+impl GitHub {
+    /* format examples:
+     * https://github.com/pjhades/nvim-repolink/blob/master/src/lib.rs
+     * https://github.com/psyomn/music/blob/feature/faim-ost/faim-ost/main-theme.ly
+     * https://github.com/psyomn/zig-getopt/blob/v1.0.1-fake/getopt.zig */
+    pub const HOST: &'static str = "github.com";
+    pub fn project_url(d: &Data) -> String {
+        let project = d.project;
+        let owner = d.owner;
+        let host = GitHub::HOST;
+        format!("https://{host}/{owner}/{project}")
+    }
+
+    pub fn service_path(d: &Data) -> String {
```

Comment:
I'm wondering if introducing a trait would be a better idea, like

```rust
trait GitHostingSite {
    fn project_url(data: &Data) -> String;
    fn service_path(data: &Data) -> String;
}

struct GitHub;

impl GitHostingSite for GitHub {
    // ...
}

pub struct Data<T: GitHostingSite> {
    hosting_site: T,
    pub project: &'a str,
    // ...
}

impl<T: GitHostingSite> Display for Data<T> {
    fn (&self, f: &mut Formatter<'_>) -> Result<(), Error> {
        // use self.hosting_site.project_url(self)
    }
}
```
Maybe `Data` should be renamed as something more meaningful. Anyway I can merge this PR and refactor it myself.